### PR TITLE
Dev shell improvements 

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

These are good in general, but in particular (except for the code motion) are needed before we drop the old make build system.

# Context

#2503 

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -294,109 +294,24 @@
...
-          buildInputs = attrs.buildInputs or []
-            ++ [
-              pkgs.gtest
-              pkgs.rapidcheck
-            ]
-            ++ lib.optional havePerl pkgs.perl
-            ;
-        });
-        in
+        makeShell = import ./packaging/dev-shell.nix { inherit lib devFlake; };
```
Comment (roberth) [inline flake.nix:297]: `callPackage` might be nice, but also a bigger change.

File: flake.nix [hunk 2]
```diff
@@ -294,109 +294,24 @@
...
+        prefixAttrs = prefix: lib.mapAttrs' (k: v: lib.nameValuePair "${prefix}-${k}" v);
```
Comment (roberth) [inline flake.nix:298]: I can't get over how universal `concatMapAttrs` is, and I admit this isn't necessary:

```suggestion
        prefixAttrs = prefix: lib.concatMapAttrs (k: v: { "${prefix}-${k}" = v; });
```

File: packaging/dev-shell.nix [hunk 3]
```diff
@@ -1,10 +1,11 @@
 { lib, devFlake }:
 
-pkgs: stdenv:
+{ pkgs }:
 
 (pkgs.nix.override { forDevShell = true; }).overrideAttrs (attrs:
 
 let
+  stdenv = pkgs.nixDependencies.stdenv;
```
Comment (roberth) [inline packaging/dev-shell.nix:8]: If the shell were part of `nixComponents`, this would just be a `stdenv` parameter and it all works out.

**Review by roberth — APPROVED** [review]

**Comment by Ericson2314** [comment]
Applied the one suggestion, happy to keep reworking things re the others. Just wanted to do enough to unblock the full Meson switchover PR.

**Pull request merged**